waspc/src/Wasp/Lib.hs

```diff
@@ -37,64 +39,74 @@ compile ::
-        validationErrors -> do
-          return ([], map show validationErrors)
+        validationErrors -> return ([], map show validationErrors)
   return $ (analyzerWarnings, []) <> compilerWarningsAndErrors
 
 analyzeWaspProject ::
   Path' Abs (Dir WaspProjectDir) ->
   CompileOptions ->
-  IO ([CompileWarning], Either (NonEmpty CompileError) AS.AppSpec)
-analyzeWaspProject waspDir options = do
-  maybeWaspFilePath <- findWaspFile waspDir
-  appSpecOrAnalyzerErrors <- case maybeWaspFilePath of
-    Nothing -> return $ Left $ fromList ["Couldn't find a single *.wasp file."]
-    Just waspFilePath -> do
-      waspFileContent <- readFile (SP.fromAbsFile waspFilePath)
-      case Analyzer.analyze waspFileContent of
-        Left analyzeError ->
-          return $
-            Left $
-              fromList
-                [ showCompilerErrorForTerminal
-                    (waspFilePath, waspFileContent)
-                    (getErrorMessageAndCtx analyzeError)
-                ]
-        Right decls -> do
-          externalCodeFiles <-
-            ExternalCode.readFiles (CompileOptions.externalCodeDirPath options)
-          maybeDotEnvServerFile <- findDotEnvServer waspDir
-          maybeDotEnvClientFile <- findDotEnvClient waspDir
-          maybeMigrationsDir <- findMigrationsDir waspDir
-          return $
-            Right
-              AS.AppSpec
-                { AS.decls = decls,
-                  AS.externalCodeFiles = externalCodeFiles,
-                  AS.externalCodeDirPath = CompileOptions.externalCodeDirPath options,
-                  AS.migrationsDir = maybeMigrationsDir,
-                  AS.dotEnvServerFile = maybeDotEnvServerFile,
-                  AS.dotEnvClientFile = maybeDotEnvClientFile,
-                  AS.isBuild = CompileOptions.isBuild options
-                }
-  analyzerWarnings <- warnIfDotEnvPresent waspDir
-  return (analyzerWarnings, appSpecOrAnalyzerErrors)
+  IO (Either (NonEmpty CompileError) AppSpec)
+analyzeWaspProject waspDir options = runExceptT $ do
+  waspFilePath <- ExceptT $ findWaspFilePath waspDir
+  declarations <- ExceptT $ analyzeWaspFileContent waspFilePath
+  liftIO $ createAppSpec waspDir options declarations
 
 -- | Checks the wasp directory for potential problems, and issues warnings if any are found.
 warnIfDotEnvPresent :: Path' Abs (Dir WaspProjectDir) -> IO [CompileWarning]
 warnIfDotEnvPresent waspDir = do
   maybeDotEnvFile <- findDotEnv waspDir
-  case maybeDotEnvFile of
-    Nothing -> return []
-    Just _ -> return ["Wasp .env files should be named .env.server or .env.client, depending on their use."]
+  return $ case maybeDotEnvFile of
+    Nothing -> []
+    Just _ -> ["Wasp .env files should be named .env.server or .env.client, depending on their use."]
+
+findWaspFilePath :: Path' Abs (Dir WaspProjectDir) -> IO (Either (NonEmpty CompileError) (Path' Abs File'))
```

Comment:
A bit weird that this function returns compile error, it doesn't have anything to do with compilation.

So we know this function either finds a file, or it doesn't find it. Therefore, the appropriate type to return is `IO (Maybe (Path' Abs File'))`.

Reply:
Aha, I see now that you just wrote this as a wrapper around `findWaspFile`, which returns Maybe.

Ha, I get what you want to do, to extract this piece, but it doesn't feel completely right. You are using the same name + `Path` -> it still feels like the same name, so it is hard to differentiate between these two functions now, and there's no good reason for one of them having a Path suffix and another one not having it.

So what are you trying to do? You are trying to add error message to it, in case when file is not found. Therefore, I would name this `findWaspFileOrGetErrorMessage`.
But that is still long. The question arises, does this even have to be a function? Maybe it is just not meant to be a function.

Reply:
Yeah, I finished reading the file, I would just move this one back to where it is called, possibly into `where`, it doesn't have enough substance to stand alone.

Reply:
I changed this entire thing, hopefully for the better. The function was only used in two places and they both needed it to return an `Either` instead of a `Maybe` (i.e., we were prepending the same error text on both call sites, so I moved the common functionality into the function). It now returns an `Either` value.

Reply:
Got it! One solution could be to still have it return Maybe, but then have somewhere else defined the message to show when that file is missing, removing duplication in that way while still keeping the function "lean" in the sense that it focuses on returning file or letting you know there is none (via Maybe) instead of also taking care of figuring out how to represent it to the user. `Maybe File` as a return type gives us more info then `Either String File` in the sense we know we get file or we don't, while in the second one we don't know how many different errors there might happen.

But for this small use case it is ok I think, they way you did it, doesn't matter that much.

Reply:
I agree and wanted to do something like this, but then I'd again need a function with a name like `findWaspFileOrGetErrorMessage` (and that would be the only call site of `findWaspFile`), so I thought this compromise was good enough. We can always refactor it if we end up needing the function somewhere else.